.circleci/config.yml

```diff
@@ -59,22 +52,37 @@ jobs:
       - run:
           name: Nodejs Version
           command: node --version
-      - *restore_yarn_cache
-      - *run_yarn
+      - restore_cache:
+          name: Restore yarn cache
+          key: v2-node-{{ arch }}-{{ checksum "yarn.lock" }}-yarn
+      - run:
+          name: Install Packages
+          command: yarn --frozen-lockfile --cache-folder ~/.cache/yarn
       - save_cache:
-          name: Save node_modules cache
-          key: v2-node-{{ arch }}-{{ .Branch }}-{{ checksum "yarn.lock" }}
+          # Store the yarn cache globally for all lock files with this same
+          # checksum. This will speed up the setup job for all PRs where the
+          # lockfile is the same.
+          name: Save yarn cache for future installs
+          key: v2-node-{{ arch }}-{{ checksum "yarn.lock" }}-yarn
           paths:
             - ~/.cache/yarn
+      - save_cache:
+          # Store node_modules for all jobs in this workflow so that they don't
+          # need to each run a yarn install for each job. This will speed up
+          # all jobs run on this branch with the same lockfile.
+          name: Save node_modules cache
+          # This cache key is per branch, a yarn install in setup is required.
+          key: v2-node-{{ arch }}-{{ .Branch }}-{{ checksum "yarn.lock" }}-node-modules
+          paths:
+            - node_modules
```

Comment:
We're now saving two different caches:
- **yarn**: This is the global yarn node_module cache to be used only by the `setup` job to speed up `yarn install` across all jobs and PRs with the same `yarn.lock`. It's only restored in the setup job.
- **node_modules**: Once the `setup` job installs a node_modules directory, we cache `node_modules` to use in every child job. This means that we only need to run `yarn install` once in the `setup` and never for the child jobs. 

The `node_modules` cache key is per branch just so that it's scoped per PR. Caches in circle are global and immutable so this just prevents unexpected collisions and should not change performance.

Reply:
I _think_ this needs to be modified to also save nested `node_modules` paths, to avoid causing differences on CI when there are dependency differences between modules (e.g. like the failing tests on #21641).